File: cosc3320/assignment1/q1/hanoi_graph.py
#!/usr/bin/env python3
"""
COSC 3320 - Assignment 1, Question 1
Towers of Hanoi on the graph G = (V, E) with

    V = {Start, A1, A2, A3, A4, Dest}
    E = {(Start,A1), (A1,A2), (A2,A3), (A3,A4), (A4,A1), (A1,Dest)}

A disk may be moved from peg u to peg v only if (u,v) is an edge of G and the
usual Hanoi rule holds (v is empty, or the top disk of v is larger).

Algorithm -- graph-aware Frame-Stewart recursion.  See README.md for the
derivation, the correctness invariant and the complexity analysis.

    TRANSFER(k, src, dst, A):        A = the set of pegs that may be used
      k == 0 : nothing
      k == 1 : walk the disk one edge at a time along a shortest src->dst
               path inside the induced subgraph G[A]
      k >= 2 : pick a spare peg x in A\\{src,dst} and a split 1 <= m < k:
                 TRANSFER(m,     src, x,   A)        park m smallest on x
                 TRANSFER(k - m, src, dst, A\\{x})    move k-m largest
                 TRANSFER(m,     x,   dst, A)        put m smallest back

x and m are chosen by an exact dynamic program over (k, src, dst, A).  Because
|V| = 6 there are only 2**6 = 64 possible peg sets.

Usage:  python3 hanoi_graph.py        then type the number of disks when asked
        echo "1 2 3 4 5 6 7 8 9 10 0" | python3 hanoi_graph.py   (part b in one go)
"""

import sys
from collections import deque

PEG_NAME = ["Start", "A1", "A2", "A3", "A4", "Dest"]
START, A1, A2, A3, A4, DEST = range(6)
EDGES = [(START, A1), (A1, A2), (A2, A3), (A3, A4), (A4, A1), (A1, DEST)]

NPEG = len(PEG_NAME)
FULL = (1 << NPEG) - 1            # the peg set {Start, A1, A2, A3, A4, Dest}
HEAD = 100                        # moves printed at the head of a long list
TAIL = 100                        # moves printed at the tail of a long list

# exact optima from hanoi_bfs.py, for comparison only
OPTIMUM = {1: 2, 2: 6, 3: 10, 4: 16, 5: 22, 6: 28,
           7: 36, 8: 44, 9: 52, 10: 62, 11: 74}


# --------------------------------------------------------------------- graph

def build_adjacency():
    adj = [set() for _ in range(NPEG)]
    for u, v in EDGES:
        adj[u].add(v)
        adj[v].add(u)
    return adj


ADJ = build_adjacency()


def all_pairs():
    """dist[A][s][t] = length of a shortest s->t path using only pegs of A.
       hop [A][s][t] = the first peg after s on such a path (None if none)."""
    dist = [[[None] * NPEG for _ in range(NPEG)] for _ in range(FULL + 1)]
    hop = [[[None] * NPEG for _ in range(NPEG)] for _ in range(FULL + 1)]
    for A in range(FULL + 1):
        for s in range(NPEG):
            if not A >> s & 1:
                continue
            dist[A][s][s] = 0
            queue = deque([s])
            while queue:                                   # plain BFS in G[A]
                u = queue.popleft()
                for v in ADJ[u]:
                    if A >> v & 1 and dist[A][s][v] is None:
                        dist[A][s][v] = dist[A][s][u] + 1
                        hop[A][s][v] = v if u == s else hop[A][s][u]
                        queue.append(v)
    return dist, hop


DIST, HOP = all_pairs()


# ----------------------------------------------------------- dynamic program

def plan(n):
    """cost[k][A][s][t] -> (moves, spare peg x, split m); the Frame-Stewart DP."""
    cost = [[[[None] * NPEG for _ in range(NPEG)]
             for _ in range(FULL + 1)] for _ in range(n + 1)]

    for A in range(FULL + 1):
        for s in range(NPEG):
            for t in range(NPEG):
                cost[0][A][s][t] = (0, None, None)
                if n >= 1:
                    d = 0 if s == t else DIST[A][s][t]
                    cost[1][A][s][t] = (d, None, None) if d is not None else None

    for k in range(2, n + 1):
        for A in range(FULL + 1):
            for s in range(NPEG):
                if not A >> s & 1:
                    continue
                for t in range(NPEG):
                    if not A >> t & 1:
                        continue
                    if s == t:
                        cost[k][A][s][t] = (0, None, None)
                        continue
                    best = None
                    for x in range(NPEG):
                        if not A >> x & 1 or x in (s, t):
                            continue
                        B = A & ~(1 << x)
                        if DIST[B][s][t] is None:       # x is a cut vertex of G[A]
                            continue
                        for m in range(1, k):
                            parts = (cost[m][A][s][x],
                                     cost[k - m][B][s][t],
                                     cost[m][A][x][t])
                            if any(p is None for p in parts):
                                continue
                            total = sum(p[0] for p in parts)
                            if best is None or total < best[0]:
                                best = (total, x, m)
                    cost[k][A][s][t] = best
    return cost


# ------------------------------------------------------------------- solving

class Board:
    """The six pegs, with every move checked against the rules of the game."""

    def __init__(self, n):
        self.n = n
        self.peg = [[] for _ in range(NPEG)]       # bottom first
        self.peg[START] = list(range(n, 0, -1))    # disk n at the bottom

    def move(self, src, dst):
        if not self.peg[src]:
            raise ValueError(f"{PEG_NAME[src]} is empty")
        disk = self.peg[src][-1]
        if dst not in ADJ[src]:
            raise ValueError(f"({PEG_NAME[src]},{PEG_NAME[dst]}) is not an edge of G")
        if self.peg[dst] and self.peg[dst][-1] < disk:
            raise ValueError(f"disk {disk} onto smaller disk {self.peg[dst][-1]} "
                             f"on {PEG_NAME[dst]}")
        self.peg[dst].append(self.peg[src].pop())
        return disk, src, dst

    def solved(self):
        return (self.peg[DEST] == list(range(self.n, 0, -1))
                and all(not self.peg[p] for p in range(NPEG) if p != DEST))


def transfer(board, cost, k, s, t, A):
    """Generator of legal moves that carries the top k disks from s to t."""
    if k <= 0 or s == t:
        return
    if k == 1:                                     # walk it edge by edge
        cur = s
        while cur != t:
            nxt = HOP[A][cur][t]
            yield board.move(cur, nxt)
            cur = nxt
        return
    entry = cost[k][A][s][t]
    if entry is None:
        raise ValueError(f"no plan for k={k} {PEG_NAME[s]}->{PEG_NAME[t]}")
    _, x, m = entry
    yield from transfer(board, cost, m,     s, x, A)
    yield from transfer(board, cost, k - m, s, t, A & ~(1 << x))
    yield from transfer(board, cost, m,     x, t, A)


def fmt(no, move):
    disk, src, dst = move
    return f"{no:10d}.  disk {disk:<3d}  {PEG_NAME[src]:<5s} -> {PEG_NAME[dst]:<5s}"


def run(n, cost, quiet=False):
    total = cost[n][FULL][START][DEST][0]
    board = Board(n)

    print()
    print("=" * 63)
    line = f" n = {n}   moves = {total}   (lower bound 2n = {2 * n}"
    if n in OPTIMUM:
        line += f",  optimum = {OPTIMUM[n]}"
    print(line + ")")
    print("=" * 63)

    tail = deque(maxlen=TAIL)
    count = 0
    for move in transfer(board, cost, n, START, DEST, FULL):
        count += 1
        if quiet:
            continue
        if total <= HEAD + TAIL or count <= HEAD:
            print(fmt(count, move))
        else:
            tail.append((count, move))

    if not quiet and total > HEAD + TAIL:
        print(f"   ... {total - HEAD - TAIL} moves omitted ...")
        for no, move in tail:
            print(fmt(no, move))

    # the algorithm is only trusted once the board agrees
    assert count == total, f"emitted {count} moves, planned {total}"
    assert board.solved(), "disks are not stacked in order on Dest"
    print(f"  [verified: {count} legal moves, "
          f"all {n} disks stacked in order on Dest]")
    return total


_pending = deque()                        # numbers already typed on one line


def read_n():
    """Read one number of disks from standard input; None means 'stop'.

    Several numbers may be given on the same line, so both
        1 <enter> 2 <enter> 0 <enter>
    and
        1 2 0 <enter>
    work, the way C++'s `cin >> n` does."""
    while True:
        if not _pending:
            print("\nEnter the number of disks n (0 to quit): ", end="", flush=True)
            try:
                line = input()
            except EOFError:                  # end of input
                print()
                return None
            _pending.extend(line.split())
            if not _pending:                  # a blank line: ask again
                continue
        word = _pending.popleft()
        try:
            n = int(word)
        except ValueError:
            print(f"  '{word}' is not a whole number.")
            continue
        if n == 0:
            return None
        if n < 0:
            print("  n must be positive.")
            continue
        if n > 60:
            print("  n is too large; please use n <= 60.")
            continue
        return n


def main():
    print("Towers of Hanoi on G = (V,E)")
    print("  V = {Start, A1, A2, A3, A4, Dest}")
    print("  E = {(Start,A1), (A1,A2), (A2,A3), (A3,A4), (A4,A1), (A1,Dest)}")

    cost = None
    planned = 0
    done = []

    while True:
        n = read_n()
        if n is None:
            break
        if n > planned:                       # (re)build the DP table when needed
            cost = plan(n)
            planned = n
        done.append((n, run(n, cost)))

    if done:
        print()
        print("-" * 62)
        print("  n    moves   2n (lower bound)   optimum   3^n-1 (path only)")
        print("-" * 62)
        for n, moves in done:
            opt = OPTIMUM.get(n, "?")
            print(f"{n:3d} {moves:8d} {2 * n:12d} {opt:>15} {3 ** n - 1:17d}")


if __name__ == "__main__":
    main()

File: cosc3320/assignment1/q1/hanoi_graph.c
/*
 * COSC 3320 - Assignment 1, Question 1
 * Towers of Hanoi on the graph G = (V,E) with
 *     V = { Start, A1, A2, A3, A4, Dest }
 *     E = { (Start,A1), (A1,A2), (A2,A3), (A3,A4), (A4,A1), (A1,Dest) }
 *
 * A disk may be moved from peg u to peg v only if (u,v) is an edge of G and the
 * usual Hanoi rule holds (v is empty, or the top disk of v is larger).
 *
 * Algorithm: graph-aware Frame-Stewart recursion.  See README.md for the
 * derivation, the correctness invariant and the complexity analysis.
 *
 *     TRANSFER(k, src, dst, A):        A = set of pegs that may be used
 *       k = 0 : nothing
 *       k = 1 : walk the disk along a shortest src->dst path inside G[A]
 *       k > 1 : pick spare x in A\{src,dst} and split 1 <= m < k, then
 *                 TRANSFER(m,   src, x,   A)
 *                 TRANSFER(k-m, src, dst, A\{x})
 *                 TRANSFER(m,   x,   dst, A)
 *
 * x and m are chosen by an exact dynamic program over (k, src, dst, A); since
 * |V| = 6 there are only 2^6 = 64 possible peg sets.
 *
 * Build: gcc -O2 -Wall -o hanoi_graph hanoi_graph.c
 * Usage: ./hanoi_graph            solve n = 1..10
 *        ./hanoi_graph N          solve n = N
 *        ./hanoi_graph MIN MAX    solve n = MIN..MAX
 *        ./hanoi_graph -s MIN MAX move counts only, no move list
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define NPEG 6
#define FULL ((1 << NPEG) - 1)
#define INF  (~0ULL >> 1)
#define HEAD 100                 /* moves printed at the head of a long list */
#define TAIL 100                 /* moves printed at the tail of a long list */

enum { START = 0, A1 = 1, A2 = 2, A3 = 3, A4 = 4, DEST = 5 };
static const char *NAME[NPEG] = { "Start", "A1", "A2", "A3", "A4", "Dest" };

/* ------------------------------------------------------------------ graph - */

static int adj[NPEG][NPEG];

static void build_graph(void)
{
    static const int E[][2] = {
        { START, A1 }, { A1, A2 }, { A2, A3 }, { A3, A4 }, { A4, A1 }, { A1, DEST }
    };
    memset(adj, 0, sizeof adj);
    for (size_t i = 0; i < sizeof E / sizeof E[0]; i++) {
        adj[E[i][0]][E[i][1]] = 1;
        adj[E[i][1]][E[i][0]] = 1;
    }
}

/* dist[A][s][t] : length of a shortest s->t path using only pegs of the set A.
 * hop [A][s][t] : the first peg after s on such a path.                      */
static int dist[1 << NPEG][NPEG][NPEG];
static int hop [1 << NPEG][NPEG][NPEG];

static void all_pairs(void)
{
    for (int A = 0; A <= FULL; A++)
        for (int s = 0; s < NPEG; s++) {
            for (int t = 0; t < NPEG; t++) { dist[A][s][t] = -1; hop[A][s][t] = -1; }
            if (!(A >> s & 1)) continue;
            int q[NPEG], qh = 0, qt = 0;                     /* plain BFS */
            dist[A][s][s] = 0; q[qt++] = s;
            while (qh < qt) {
                int u = q[qh++];
                for (int v = 0; v < NPEG; v++)
                    if (adj[u][v] && (A >> v & 1) && dist[A][s][v] < 0) {
                        dist[A][s][v] = dist[A][s][u] + 1;
                        hop [A][s][v] = (u == s) ? v : hop[A][s][u];
                        q[qt++] = v;
                    }
            }
        }
}

/* ------------------------------------------------------- dynamic program - */

static unsigned long long *cost;     /* cost[k][A][s][t] = # moves            */
static signed char        *spare;    /* the peg x chosen                      */
static int                *split;    /* the split m chosen                    */

static size_t idx(int k, int A, int s, int t)
{
    return (((size_t)k * (FULL + 1) + A) * NPEG + s) * NPEG + t;
}

static void plan(int n)
{
    size_t cells = idx(n + 1, 0, 0, 0);
    cost  = malloc(cells * sizeof *cost);
    spare = malloc(cells * sizeof *spare);
    split = malloc(cells * sizeof *split);
    if (!cost || !spare || !split) { fprintf(stderr, "out of memory\n"); exit(1); }

    for (int A = 0; A <= FULL; A++)
        for (int s = 0; s < NPEG; s++)
            for (int t = 0; t < NPEG; t++) {
                cost[idx(0, A, s, t)] = 0;
                if (n >= 1)
                    cost[idx(1, A, s, t)] =
                        (s == t) ? 0 : (dist[A][s][t] < 0 ? INF : (unsigned long long)dist[A][s][t]);
            }

    for (int k = 2; k <= n; k++)
        for (int A = 0; A <= FULL; A++)
            for (int s = 0; s < NPEG; s++) {
                if (!(A >> s & 1)) continue;
                for (int t = 0; t < NPEG; t++) {
                    if (!(A >> t & 1)) continue;
                    if (s == t) { cost[idx(k, A, s, t)] = 0; continue; }
                    unsigned long long best = INF;
                    int bx = -1, bm = -1;
                    for (int x = 0; x < NPEG; x++) {
                        if (!(A >> x & 1) || x == s || x == t) continue;
                        int B = A & ~(1 << x);
                        if (dist[B][s][t] < 0) continue;      /* x is a cut vertex */
                        for (int m = 1; m < k; m++) {
                            unsigned long long a = cost[idx(m,     A, s, x)];
                            unsigned long long b = cost[idx(k - m, B, s, t)];
                            unsigned long long c = cost[idx(m,     A, x, t)];
                            if (a == INF || b == INF || c == INF) continue;
                            if (a + b + c < best) { best = a + b + c; bx = x; bm = m; }
                        }
                    }
                    cost [idx(k, A, s, t)] = best;
                    spare[idx(k, A, s, t)] = (signed char)bx;
                    split[idx(k, A, s, t)] = bm;
                }
            }
}

/* --------------------------------------------------- board + move output - */

static int  *peg[NPEG];              /* peg[p][0..hgt[p]-1], bottom first     */
static int   hgt[NPEG];
static long long nmove;              /* moves emitted so far                  */
static int   quiet;                  /* -s : print counts only                */
static long long total;              /* total moves of the current run        */

struct rec { int disk, from, to; };
static struct rec ring[TAIL];        /* circular buffer with the last moves   */

static void line(long long no, struct rec r)
{
    printf("%10lld.  disk %-3d  %-5s -> %-5s\n", no, r.disk, NAME[r.from], NAME[r.to]);
}

static void emit(int from, int to)
{
    if (hgt[from] == 0) {
        fprintf(stderr, "ILLEGAL: %s is empty\n", NAME[from]); exit(1);
    }
    int d = peg[from][hgt[from] - 1];
    if (!adj[from][to]) {
        fprintf(stderr, "ILLEGAL: (%s,%s) is not an edge\n", NAME[from], NAME[to]); exit(1);
    }
    if (hgt[to] > 0 && peg[to][hgt[to] - 1] < d) {
        fprintf(stderr, "ILLEGAL: disk %d onto smaller disk %d on %s\n",
                d, peg[to][hgt[to] - 1], NAME[to]); exit(1);
    }
    hgt[from]--;
    peg[to][hgt[to]++] = d;
    nmove++;

    if (quiet) return;
    struct rec r = { d, from, to };
    if (total <= HEAD + TAIL || nmove <= HEAD) line(nmove, r);
    else ring[nmove % TAIL] = r;                 /* remember, print at the end */
}

/* ------------------------------------------------------------- transfer - */

static void transfer(int k, int s, int t, int A)
{
    if (k <= 0 || s == t) return;
    if (k == 1) {                                /* walk it edge by edge */
        int cur = s;
        while (cur != t) { int nx = hop[A][cur][t]; emit(cur, nx); cur = nx; }
        return;
    }
    int x = spare[idx(k, A, s, t)];
    int m = split[idx(k, A, s, t)];
    if (x < 0) { fprintf(stderr, "no plan for k=%d %s->%s\n", k, NAME[s], NAME[t]); exit(1); }
    transfer(m,     s, x, A);
    transfer(k - m, s, t, A & ~(1 << x));
    transfer(m,     x, t, A);
}

/* ------------------------------------------------------------------ main - */

static long long optimum[] = {          /* exhaustive BFS, see hanoi_bfs.c */
    0, 2, 6, 10, 16, 22, 28, 36, 44, 52, 62, 74
};
#define NOPT ((int)(sizeof optimum / sizeof optimum[0]))

static void run(int n)
{
    for (int p = 0; p < NPEG; p++) { peg[p] = malloc((size_t)n * sizeof(int)); hgt[p] = 0; }
    for (int d = n; d >= 1; d--) peg[START][hgt[START]++] = d;   /* n at the bottom */

    total = (long long)cost[idx(n, FULL, START, DEST)];
    nmove = 0;

    printf("\n===============================================================\n");
    printf(" n = %d   moves = %lld   (lower bound 2n = %d", n, total, 2 * n);
    if (n < NOPT) printf(",  optimum = %lld", optimum[n]);
    printf(")\n");
    printf("===============================================================\n");

    transfer(n, START, DEST, FULL);

    if (!quiet && total > HEAD + TAIL) {
        printf("   ... %lld moves omitted ...\n", total - HEAD - TAIL);
        for (long long i = total - TAIL + 1; i <= total; i++) line(i, ring[i % TAIL]);
    }

    /* the algorithm is only trusted once the board agrees */
    if (nmove != total) { fprintf(stderr, "move count mismatch\n"); exit(1); }
    if (hgt[DEST] != n) { fprintf(stderr, "Dest does not hold all disks\n"); exit(1); }
    for (int i = 0; i < n; i++)
        if (peg[DEST][i] != n - i) { fprintf(stderr, "Dest is not sorted\n"); exit(1); }
    for (int p = 0; p < NPEG; p++)
        if (p != DEST && hgt[p] != 0) { fprintf(stderr, "%s not empty\n", NAME[p]); exit(1); }
    printf("  [verified: %lld legal moves, all %d disks stacked in order on Dest]\n", nmove, n);

    for (int p = 0; p < NPEG; p++) free(peg[p]);
}

int main(int argc, char **argv)
{
    int lo = 1, hi = 10, a = 1;
    if (a < argc && strcmp(argv[a], "-s") == 0) { quiet = 1; a++; }
    if (a < argc) lo = hi = atoi(argv[a++]);
    if (a < argc) hi = atoi(argv[a++]);
    if (lo < 1 || hi < lo) { fprintf(stderr, "usage: %s [-s] [MIN [MAX]]\n", argv[0]); return 1; }

    build_graph();
    all_pairs();
    plan(hi);

    printf("Towers of Hanoi on G = (V,E)\n");
    printf("  V = {Start, A1, A2, A3, A4, Dest}\n");
    printf("  E = {(Start,A1), (A1,A2), (A2,A3), (A3,A4), (A4,A1), (A1,Dest)}\n");

    for (int n = lo; n <= hi; n++) run(n);

    printf("\n--------------------------------------------------------------\n");
    printf("  n    moves   2n (lower bound)   optimum   3^n-1 (path only)\n");
    printf("--------------------------------------------------------------\n");
    for (int n = lo; n <= hi; n++) {
        unsigned long long p3 = 1;
        for (int i = 0; i < n; i++) p3 *= 3;
        printf("%3d %8llu %12d ", n, cost[idx(n, FULL, START, DEST)], 2 * n);
        if (n < NOPT) printf("%15lld ", optimum[n]); else printf("%15s ", "?");
        printf("%17llu\n", p3 - 1);
    }
    free(cost); free(spare); free(split);
    return 0;
}
